ENH: Use rfft when possible in resampling

larsoner opened the pull request:
Taking over #5610. Ready for review/merge from my end.

@endolith can you look to make sure you're (still) happy with this?

Closes #5610
Closes #5592

~~EDIT: #10361 should go in first, then this can proceed~~

endolith commented:
@larsoner The diff on github shows it removing optimizations to fftconvolve. Is that real?

larsoner commented:
> The diff on github shows it removing optimizations to fftconvolve

I don't see this happening in the diff, where?

larsoner commented:
@endolith any time to look in the next few days? If so this might make 1.3.0. If not we can push to 1.4.0

endolith commented on scipy/signal/signaltools.py:
```diff
@@ -408,24 +401,13 @@ def fftconvolve(in1, in2, mode="full", axes=None):
-        finally:
-            if not _rfft_mt_safe:
-                _rfft_lock.release()
+    if not complex_result:
```
@larsoner I mean these seemingly unrelated changes to `fftconvolve`.  Is this from not being rebased to the latest master?

larsoner replied:
It's related in the sense that whatever triaging / complexity we have in `resample` should look a lot like what we have in `fftconvolve`. SciPy now requires a new enough `numpy` that we don't need to worry about this locking business anymore, so after this PR it shouldn't be in either code path (`fftconvolve` or `resample`).

WarrenWeckesser replied:
Those changes should be explained in the commit message.  [Was this project dragging on?](https://xkcd.com/1296/)

larsoner replied:
Good call, done

endolith commented:
I tried to use `hypothesis` to compare the outputs for a variety of inputs, and the new `fftconvolve` seems fine, but it found an inconsistent case for the new `resample`:

```
resample([1], 2)
Out[20]: array([1., 1.])

scipy.signal.resample([1], 2)
Out[21]: array([2., 0.])
```

endolith commented:
So that's a change in behavior of the new algorithm:

```
resample([1], 5)
Out[93]: array([1., 1., 1., 1., 1.])

resample([1+0j], 5)
Out[94]: array([5.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j])

scipy.signal.resample([1], 5)
Out[95]: array([5., 0., 0., 0., 0.])

scipy.signal.resample([1+0j], 5)
Out[96]: array([5.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j])
```

Though the old behavior might be considered a bug?

```
scipy.signal.resample([1], 5)
Out[98]: array([5., 0., 0., 0., 0.])

scipy.signal.resample([1, 1], 5)
Out[99]: array([1., 1., 1., 1., 1.])

scipy.signal.resample([1, 1, 1], 5)
Out[100]: array([1., 1., 1., 1., 1.])
```

JesseLivezey commented on scipy/signal/signaltools.py:
```diff
@@ -2252,36 +2233,66 @@ def resample(x, num, t=None, axis=0, window=None):
-    if x.dtype.char not in ['F', 'D']:
-        y = y.real
+
+    # Can we use faster real FFT?
```
This might be more clear as "Check if we can use faster real FFT." I initially read this as a suggestion for future exploration rather than a code comment.

larsoner commented:
I think it is indeed a bug on `master` since the FFT method assumes periodicity of the signal. I've added a test that various-length DC signals resample to DC signals for both real and complex types.

FYI since I suspect @tylerjereddy will ask, benchmarking `signal.Resample` on `master` I get 600.60ms for the complex and 585.28ms for real, and on this PR I get 486.65ms and 231.02ms, respectively :)

tylerjereddy commented:
@endolith are you +1 here now?

endolith commented:
I keep trying to read through it and confirm the changes, and then get stuck and tell myself I'll finish it later, but after the *n*th attempt, I started to think "Maybe it's not me; maybe it's the code's readability".  So I started to re-order and change a few things and add some comments, etc.  but didn't finish yet

It did pass the hypothesis tests, so it's probably fine?  If you're in a rush for 1.3.0.

tylerjereddy commented:
@endolith Thanks for that. I'll bump the milestone to give you some more time. I like to give PRs the best chance I can to make it in so I do tend to ping people a little before branching.

larsoner commented:
> (Wait, can others not see these comments until I submit the complete review?)

Yes comments are in "pending" state until the review is submitted. A number of comments (at least some of the ones about `float`) are outdated because they have already been fixed. But I'll look more closely later today hopefully to address the others.

larsoner commented:
@endolith I rebased, moved comments to 1.4.0 docs, added some better benchmarks (using numbers in and out of the radix) and addressed your comments. Let me know if it looks good to you. For what it's worth, it looked like using `np.fft.rfft` was an order of magnitude faster for the prime numbers in the test I ran, so that was a good change!

larsoner commented:
@endolith okay with you now?

larsoner commented:
@toddrjen this might be relevant for you, too -- are you up for reviewing? We could use a signals expert to double check the code.

larsoner commented:
@WarrenWeckesser I'm struggling to find signal reviewers for this -- if you can spare some cycles feel free to look.

endolith commented:
My suggested changes (not sure about some of them) are in https://github.com/larsoner/scipy/pull/8

endolith commented:
Also you still need to change it to use the new scipy.fft

larsoner commented:
I was thinking of doing that in my other PR that already migrates the rest of signaltools to use scipy.fft. Either way I'll do it at some point. But I guess it shouldn't additionally complicate things too much to do it here.

larsoner commented:
> But I guess it shouldn't additionally complicate things too much to do it here.

Actually there are some complicated bits having to do with how `scipy.fft` is imported that are part of #10290 and #10361, so moving to `scipy.fft.rfft` should wait until those are merged.

In the meantime @endolith I squashed your commits after omitting one that broke the tests and restoring a `num` that was missing and breaking tests, cherry-picked it, rebased on latest `master`, and force pushed here. Let me know if you're happy, if so I think we can merge (assuming CIs are happy)!

Rebase+merge would be best here since commits have already been squashed to author-based commits.

larsoner commented:
@endolith is this good to go from your end? If so I'll go ahead and merge since this already went through a bit of review before, and since then we've just added your cleanups.

WarrenWeckesser commented:
@larsoner, I'll look at this "soon" (but not for at least a day or two).  If you and endolith (and any other reviewers) have resolved all the issues, don't wait for me.

peterbell10 commented on scipy/signal/signaltools.py:
```diff
@@ -408,24 +401,13 @@ def fftconvolve(in1, in2, mode="full", axes=None):
     fslice = tuple([slice(sz) for sz in shape])
     # Pre-1.9 NumPy FFT routines are not threadsafe.  For older NumPys, make
     # sure we only call rfftn/irfftn from one thread at a time.
```
This comment is stale.

larsoner replied:
This is fixed in #10361, I'm trying not to have too much overlap in changes

larsoner replied:
Actually maybe it's not... I'll remove it in one of these PRs at least.

endolith commented:
@larsoner I'm fine with it (I think, as long as you don't find any other mistakes in the parts I did).

larsoner commented:
Okay I've rebased after #10361 and the changes otherwise look the same.

Given multiple previous rounds of review and @endolith's +1, I'll rebase+merge this once the CIs come back happy. Thanks for doing most of the original work @dnlbunting !

endolith commented:
What are these red question marks by the way? https://github.com/scipy/scipy/pull/9876/commits/f1e598d8d60cd2f90280c97189a7fae4731d3335#diff-97a2088dd3cb8cda2ae551fa9357d5a8R2240

larsoner commented:
`rebase` conflict cruft that probably shouldn't be in the individual commits. Not worth rewriting history to fix it, though.
